cardano-api/src/Cardano/Api/MetaData.hs

```diff
@@ -78,44 +78,25 @@ validateTxMetadata
   :: TxMetadata
   -> Either TxMetadataValidationError TxMetadata
 validateTxMetadata txMd@(TxMetadata (MetaData mdMap)) =
-    Map.foldl' foldFn (Right txMd) mdMap
+    txMd <$ for_ mdMap validate
   where
-    foldFn
-      :: Either TxMetadataValidationError TxMetadata
-      -> MetaDatum
-      -> Either TxMetadataValidationError TxMetadata
-    foldFn lastResult metaDatum =
-      case lastResult of
-        Left err -> Left err
-        Right _ -> validate metaDatum
-
-    validate :: MetaDatum -> Either TxMetadataValidationError TxMetadata
+    validate :: MetaDatum -> Either TxMetadataValidationError ()
     validate metaDatum =
       case metaDatum of
-        Map [] -> Right txMd
-        Map ((k, v):mdPairs) ->
-          case (validate k, validate v) of
-            (Left err, _) -> Left err
-            (_, Left err) -> Left err
-            _ -> validate (Map mdPairs)
-
-        List [] -> Right txMd
-        List (md:mds) ->
-          case validate md of
-            Right _ -> validate (List mds)
-            Left err -> Left err
+        Map mdPairs -> for_ mdPairs $ \(k, v) -> validate k >> validate v
+        List mds -> for_ mds validate
```

Comment:
Use `for_` which is the same as `flip traverse_`, we can write this more intuitively.

Logically we are looping over the list or map and validating each element or key value and aborting on the first failure.